examples/doremi/doremi/dataloader.py

```diff
@@ -0,0 +1,363 @@
+        self.microbatch_idx = 0
+        self.domain_counters = [0 for _ in self.datasets]
+        self.total_samples_yielded = 0
+        self.out_of_samples = False
+
+        domain_indices = []
+        for i, dataset in enumerate(self.datasets):
+            local_indices = torch.arange(0, len(dataset), device="cpu").tolist()
+
+            # NOTE: align the indices across the combined dataset
+            global_indices = local_indices + self.offsets[i]
+            domain_indices.append(global_indices)
+
+        self.num_samples_per_global_step = self.batch_size * self.num_microbatches * self.num_replicas
+        self.domain_indices = domain_indices
+        self.expected_total_samples = sum([len(d) for d in domain_indices])
+
+
+def get_datasets(paths):
+    datasets = []
+    for path in tqdm(paths, desc="Loading dataset from disk"):
+        d = load_from_disk(path)
+        datasets.append(d)
+
+    return datasets
+
+
+def get_dataloader(trainer: DistributedTrainer, datasets) -> DataLoader:
+    doremi_context = trainer.doremi_context
+    parallel_context = trainer.parallel_context
+
+    datasets = [d.with_format(type="numpy", columns=["input_ids"], output_all_columns=True) for d in datasets]
+
+    # TODO(xrsrke): decouple trainer from dataloader
+    # TODO(xrsrke): decouple data collating from data loading
+    input_pp_rank, output_pp_rank = get_input_output_pp_ranks(model=trainer.model)
+    data_collator = DataCollatorForCLM(
+        sequence_length=trainer.sequence_length,
+        input_pp_rank=input_pp_rank,
+        output_pp_rank=output_pp_rank,
+        parallel_context=parallel_context,
+        doremi_context=doremi_context,
+    )
+
+    sampler = DistributedSamplerForDoReMi(
+        datasets,
+        batch_size=trainer.micro_batch_size,
+        num_microbatches=trainer.n_micro_batches_per_batch,
+        num_replicas=parallel_context.dp_pg.size(),
+        rank=dist.get_rank(parallel_context.dp_pg),
+        seed=trainer.config.data.seed,
+        drop_last=True,
+        doremi_context=doremi_context,
+        parallel_context=parallel_context,
+    )
+
+    comebined_dataset = CombinedDataset(datasets)
+
+    dataloader = DataLoader(
+        comebined_dataset,
+        batch_sampler=sampler,
+        collate_fn=data_collator,
+        num_workers=trainer.config.data.num_loading_workers,
+        pin_memory=True,
+        worker_init_fn=get_dataloader_worker_init(dp_rank=dist.get_rank(parallel_context.dp_pg)),
+    )
+
+    def _data_generator(dataloader):
+        def inner():
+            for batch in dataloader:
+                batch = {k: v.to("cuda") for k, v in batch.items()}
+                # NOTE: because the inference model don't take `domain_idxs`
+                # as input we need to remove it from the batch
+                batch_for_inference = {k: v for k, v in batch.items() if k != "domain_idxs"}
+
+                ref_losses = trainer.ref_model(**batch_for_inference)["losses"]
```

Comment:
add TODO to be able to prefetch a couple of batches using ref_model
So instead of fetch_1_global_batch -> train_batch_iter -> fetch_1_global_batch -> train_batch_iter..
We can have fetch_2_global_batch -> train_batch_iter -> train_batch_iter -> fetch_2_global_batch -> train_batch_iter..